presto-bigquery/src/main/java/com/facebook/presto/plugin/bigquery/ReadSessionCreator.java

```diff
@@ -0,0 +1,197 @@
+            BigQueryClient bigQueryClient,
+            BigQueryStorageClientFactory bigQueryStorageClientFactory)
+    {
+        this.config = config;
+        this.bigQueryClient = bigQueryClient;
+        this.bigQueryStorageClientFactory = bigQueryStorageClientFactory;
+    }
+
+    public Storage.ReadSession create(TableId table, ImmutableList<String> selectedFields, Optional<String> filter, int parallelism)
+    {
+        TableInfo tableDetails = bigQueryClient.getTable(table);
+
+        TableInfo actualTable = getActualTable(tableDetails, selectedFields, new String[] {});
+
+        try (BigQueryStorageClient bigQueryStorageClient = bigQueryStorageClientFactory.createBigQueryStorageClient()) {
+            ReadOptions.TableReadOptions.Builder readOptions = ReadOptions.TableReadOptions.newBuilder()
+                    .addAllSelectedFields(selectedFields);
+            filter.ifPresent(readOptions::setRowRestriction);
+
+            TableReferenceProto.TableReference tableReference = toTableReference(actualTable.getTableId());
+
+            Storage.ReadSession readSession = bigQueryStorageClient.createReadSession(
+                    Storage.CreateReadSessionRequest.newBuilder()
+                            .setParent("projects/" + bigQueryClient.getProjectId())
+                            .setFormat(Storage.DataFormat.AVRO)
+                            .setRequestedStreams(parallelism)
+                            .setReadOptions(readOptions)
+                            .setTableReference(tableReference)
+                            // The BALANCED sharding strategy causes the server to
+                            // assign roughly the same number of rows to each stream.
+                            .setShardingStrategy(Storage.ShardingStrategy.BALANCED)
+                            .build());
+
+            return readSession;
+        }
+    }
+
+    TableReferenceProto.TableReference toTableReference(TableId tableId)
+    {
+        return TableReferenceProto.TableReference.newBuilder()
+                .setProjectId(tableId.getProject())
+                .setDatasetId(tableId.getDataset())
+                .setTableId(tableId.getTable())
+                .build();
+    }
+
+    TableInfo getActualTable(
+            TableInfo table,
+            ImmutableList<String> requiredColumns,
+            String[] filters)
+    {
+        TableDefinition tableDefinition = table.getDefinition();
+        TableDefinition.Type tableType = tableDefinition.getType();
+        if (TableDefinition.Type.TABLE == tableType) {
+            return table;
+        }
+        if (TableDefinition.Type.VIEW == tableType) {
+            if (!config.viewsEnabled) {
+                throw new PrestoException(NOT_SUPPORTED, format(
+                        "Views are not enabled. You can enable views by setting '%s' to true. Notice additional cost may occur.",
+                        "bigquery.views-enabled"));
```

Comment:
don't think we really need a string format here, can we put the `bigquery.views-enabled` inline?

Reply:
sure